foolbox/attacks/adef_attack.py

```diff
@@ -0,0 +1,405 @@
+        fx = (logits.transpose() - logits[0, original_label]).transpose()
+
+        norm_full = 0  # norm of the vector field
+        vec_field_full = np.zeros((h, w, 2))  # the vector field
+
+        current_label = original_label
+        logging.info('Iterations finished: 0')
+        logging.info('Current labels: {} '.format(current_label))
+
+        for step in range(max_iter):
+            n += 1
+            logits, grad, is_adv = a.predictions_and_gradient(perturbed)
+            if is_adv:
+                a.predictions(perturbed)
+                self.vector_field = vec_field_full
+                return
+
+            d1x, d2x = _difference_map(perturbed, color_axis)
+
+            logits_for_grad = np.zeros_like(logits)
+            logits_for_grad[original_label[0]] = 1
+
+            grad_original = a.backward(logits_for_grad, perturbed)
+
+            # Find vector fields for the image and each candidate label.
+            # Keep the smallest vector field for each image.
+            norm_min = np.inf
+
+            # iterate over all candidate classes
+            for target_no in range(1, num_classes):
+
+                target_labels = candidates[0, target_no]
+                logits_for_grad = np.zeros_like(logits)
+                logits_for_grad[target_labels] = 1
+
+                # gradient of the target label w.r.t. image
+                grad_target = a.backward(logits_for_grad, perturbed)
+
+                # Derivative of the binary classifier 'F_lab - F_orig'
+                dfx = grad_target - grad_original
+
+                f_im = fx[0, candidates[0, target_no]]
+
+                # create the vector field
+                vec_field_target = _create_vec_field(
+                    f_im, dfx, d1x, d2x, color_axis, smooth
+                )
+
+                vec_field_target += vec_field_full
+
+                # l2 norm of vector field.
+                norm_target = np.linalg.norm(vec_field_target.ravel())
+
+                # choose the vector field with the smallest norm
+                if norm_target < norm_min:
+                    norm_min = norm_target
+                    vec_field_min = vec_field_target
+
+            # Update the image by applying the vector field,
+            # the vector field is always applied to the original image,
+            # since the current vector field is added to all prior
+            # vector fields via vec_field_target += vec_field_full
+            perturbed = _compose(image_original.copy(), vec_field_min,
+                                 color_axis)
+
+            vec_field_full = vec_field_min
+            norm_full = norm_min
+
+            # getting the current label after applying the vector field
+            fx, _ = a.predictions(perturbed)
+            fx = np.expand_dims(fx, axis=0)
+            current_label = np.argmax(fx, axis=1)
+            fx = (fx.transpose() - fx[0, current_label]).transpose()
+
+            # See if we have been successful.
```

Comment:
Whether you have found a valid adversarial according to the criterion (be in targeted or untargeted) is something you get back when calling `a.predictions` in the form of the `is_adv` / `is_adversarial` return value. Along with getting rid of the `targeting` argument this code should be changed to to just use the return value of `a.predictions` rather than trying to determine by itself whether an adversarial has been found or not.